File: src/main/java/com/oycm/algorithm/f/prefix_sum_hash/Solution_12.scala
package com.oycm.algorithm.f.prefix_sum_hash

object Solution_12 {

  /**
   * 1124. 表现良好的最长时间段 1908
   * https://leetcode.cn/problems/longest-well-performing-interval/description/
   *
   * hours(i) > 8 时，这一天是劳累的，否则是不劳累的
   *
   * 表现良好时间段：劳累的天数 严格大于 不劳累的天数
   *
   * @param hours 工作时间表数组
   * @return
   */
  def longestWPI(hours: Array[Int]): Int = {
    /*
    是不是可以把 hours(i) > 8 看作 1，hours(i) <= 8 的看作 -1
    sums(r) - sums(l) > 0 的最大长度
    怎么找最大长度？
    如果 sums(r) 大于 0，则 ans = r + 1
    sums(r) <= 0 时，只需要找到 sums(r) - 1 = sums(l) 第一次出现的 index，max(r - index + 1, ans), 不存在，说明没有符合要求的时间段
    如果 sums(r) == 0, [0, r-1] 之间肯定有 nums(i) == 1
    如果 nums(r) < 0, [0, r-1] 之间可能有 nums(i) == 1
     */
    var ans = 0
    var sum = 0
    val map = scala.collection.mutable.Map[Int, Int]()
    for (i <- hours.indices) {
      val num = if (hours(i) > 8) 1 else -1
      sum += num
      if (sum > 0) {
        ans = i + 1
      } else {
        val l = map.getOrElse(sum - 1, i + 1)
        ans = Math.max(ans, i - l)
      }
      map(sum) = map.getOrElse(sum, i)

    }

    ans
  }

  def optimize(hours: Array[Int]): Int = {
    /*
    对 sums 取反，
     */
    val n = hours.length
    var ans = 0
    var sum = 0
    // 数组记录前缀和首次出现的位置
    val map = Array.fill(n + 2)(n)
    for (i <- hours.indices) {
      val num = if (hours(i) > 8) -1 else 1
      sum += num
      if (sum < 0) {
        ans = i + 1
      } else {
        if (map(sum + 1) < n) {
          ans = Math.max(ans, i - map(sum + 1))
        }
        if (map(sum) == n) {
          map(sum) = i
        }
      }
    }

    ans
  }


  def main(args: Array[String]): Unit = {
    println(longestWPI(Array(9, 9, 6, 0, 6, 6, 9)))
    println(longestWPI(Array(6, 6, 6)))
    println(longestWPI(Array(6, 9, 6)))

    println(optimize(Array(9, 9, 6, 0, 6, 6, 9)))
    println(optimize(Array(6, 6, 6)))
    println(optimize(Array(6, 9, 6)))
  }

}

File: src/main/java/com/oycm/month2026/august/Solution_12.java
package com.oycm.month2026.august;

import java.util.HashMap;
import java.util.Map;

public class Solution_12 {

    /**
     * 2958. <a href="https://leetcode.cn/problems/length-of-longest-subarray-with-at-most-k-frequency/description/">最多 K 个重复元素的最长子数组</a> 1535
     *
     * @param nums
     * @param k
     * @return
     */
    public int maxSubarrayLength(int[] nums, int k) {
        /*
        给你一个整数数组 nums 和一个整数 k 。
        一个元素 x 在数组中的 频率 指的是它在数组中的出现次数。
        如果一个数组中所有元素的频率都 小于等于 k ，那么我们称这个数组是 好 数组。
        请你返回 nums 中 最长好 子数组的长度。
        子数组 指的是一个数组中一段连续非空的元素序列。
         */
        /*
        不定长滑动窗口求最长
         */
        int ans = 0;
        int l = 0;
        Map<Integer, Integer> cnt = new HashMap<>();
        for (int r = 0; r < nums.length; r++) {
            cnt.merge(nums[r], 1, Integer::sum);
            while (cnt.get(nums[r]) > k) {
                cnt.merge(nums[l], -1, Integer::sum);
                l++;
            }
            ans = Math.max(ans, r - l + 1);
        }
        return ans;
    }

}
